### PR TITLE
fix: Typescript issues in preparation for pnpm 7 upgrade

### PR DESCRIPTION
**Pull request by rjchow**
One of the blockers for PNPM 7 upgrade is the change in `public-hoist-pattern` as described in #33685.

This pull request addresses the missing Typescript definition packages across the packages in the project as well as WooCommerce core.

### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Most of the changes are just additions of `@types/*`

It doesn't seem like `eslint` will be an issue as we run eslint through our own `@woocommerce/eslint-plugin` in the workspace.

### How to test the changes in this Pull Request:

As this PR does not complete the PNPM 7 migration, in order to test this you will need to install PNPM 7 and update the 'engines' field in `plugins/woocommerce-admin/package.json`, `plugins/woocommerce/package.json` and `tools/cherry-pick/package.json` to allow PNPM 7.

Aside from that, you will also need to install PNPM 7: `npm install -g pnpm@7.8.0`

Once those steps have been completed,

1. `pnpm install`
2. `pnpm -- turbo run --no-cache build --filter=woocommerce`

With PNPM v7 you will see the error below, which can be fixed by removing the two `// @ts-expect-error` comments. 

Though I'm not entirely sure why the error goes away with PNPM v7, it doesn't seem quite related and will require a migration of the component to TS to resolve properly.

image

Once these errors are removed there should be 0 errors in the build process.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+auto-install-peers=true
```
Comment (chihsuan) [inline .npmrc:2]: @rjchow Could you add a comment to explain why we need this? Thank you!

>  Lines in .npmrc files are interpreted as comments when they begin with a ; or # character. .npmrc files are parsed by [npm/ini](https://github.com/npm/ini), which specifies this comment syntax.

File: packages/js/components/package.json [hunk 2]
```diff
@@ -107,6 +109,7 @@
 		"jest-cli": "^27.5.1",
 		"postcss-loader": "^3.0.0",
 		"require-turbo": "workspace:*",
+		"react": "^17.0.0",
```
Comment (chihsuan) [inline packages/js/components/package.json:111]: Why do we need to add "react" here? I think we have also included it in `peerDependencies`. ❓

File: plugins/woocommerce-admin/client/homescreen/layout.js [hunk 3]
```diff
@@ -67,7 +67,7 @@ export const Layout = ( {
 	const hasTwoColumnContent =
 		shouldShowStoreLinks || window.wcAdminFeatures.analytics;
 	const [ showInbox, setShowInbox ] = useState( true );
-	const isDashboardShown = ! query.task;
+	const isDashboardShown = ! query.task; // ?&task=<x> query param is used to show tasks instead of the homescreen
```
Comment (chihsuan) [inline plugins/woocommerce-admin/client/homescreen/layout.js]: Thanks for adding this helpful comment! 👍

File: packages/js/components/package.json (hunk 2 above)
Comment (rjchow) [reply packages/js/components/package.json:111]: I had to look this up to understand why: https://blog.bitsrc.io/understanding-peer-dependencies-in-javascript-dbdb4ab5a7be

it looks like we use some React stuff directly so peerDeps isn't sufficient

Comment (chihsuan) [reply packages/js/components/package.json:111]: Got it! Thanks for looking into it.

Comment (rjchow) [reply packages/js/components/package.json:111]: done!

File: .npmrc (hunk 1 above)
Comment (rjchow) [reply .npmrc:2]: *done!

**Review by chihsuan — APPROVED** [review]
@rjchow Thanks for working on this! Tested well with the instructions. 👍

**Pull request merged**